### PR TITLE
Add Fido search results to the database

### PR DESCRIPTION
**Pull request by Punyaslok**
Now Fido search results (except the ones from JSOC) can be added to the database. Each QueryResponseBlock forms a DatabaseEntry.

@Cadair

### CODE REVIEW

File: sunpy/database/tables.py [hunk 1]
```diff
@@ -429,6 +454,23 @@ def entries_from_query_result(qr, default_waveunit=None):
         yield DatabaseEntry._from_query_result_block(block, default_waveunit)
 
 
+def entries_from_fido_search_result(sr, default_waveunit=None):
+    for entry in sr:
+        if isinstance(entry, sunpy.net.vso.vso.QueryResponse):
+            #This is because EVE doesn't return a Fido QueryResponse. It
```
Comment (Cadair) [inline sunpy/database/tables.py]: This comment is misleading, it's just because Fido can query the VSO.

File: sunpy/database/database.py [hunk 2]
```diff
@@ -780,6 +780,29 @@ def add_from_vso_query_result(self, query_result,
                 query_result, self.default_waveunit),
             ignore_already_added)
 
+    def add_from_fido_search_result(self, search_result,
+            ignore_already_added=False):
+        """Generate database entries from a Fido search result and add all the
```
Comment (Cadair) [inline sunpy/database/database.py]: line break after the `"""`

File: sunpy/database/database.py [hunk 3]
```diff
@@ -785,6 +785,29 @@ def add_from_vso_query_result(self, query_result,
                 query_result, self.default_waveunit),
             ignore_already_added)
 
+    def add_from_fido_search_result(self, search_result,
+            ignore_already_added=False):
```
Comment (Cadair) [inline sunpy/database/database.py]: indentation is wrong.

File: sunpy/database/database.py [hunk 4]
```diff
@@ -785,6 +785,29 @@ def add_from_vso_query_result(self, query_result,
...
+        """Generate database entries from a Fido search result and add all the
+        generated entries to this database.
+
+        Parameters
+        ----------
+        search_result : sunpy.net.dataretriever.downloader_factory.UnifiedResponse
```
Comment (Cadair) [inline sunpy/database/database.py]: type should be in backticks.

File: sunpy/database/database.py [hunk 5]
```diff
@@ -785,6 +785,29 @@ def add_from_vso_query_result(self, query_result,
...
+            A UnifiedResponse object that is used to store responses from the
+            unified downloader. This is returned by the ``search`` method of a
+            :class:`sunpy.net.dataretriever.downloader_factory.UnifiedDownloaderFactory`
+            object.
+
+        ignore_already_added : bool
```
Comment (Cadair) [inline sunpy/database/database.py]: type should be in backticks

File: sunpy/database/tables.py [hunk 6]
```diff
@@ -340,6 +341,66 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
             instrument=instrument, size=qr_block.size,
             wavemin=wavemin, wavemax=wavemax)
 
+    @classmethod
+    def _from_fido_search_result_block(cls, sr_block, default_waveunit=None):
+        """Make a new :class:`DatabaseEntry` instance from a Fido search
```
Comment (Cadair) [inline sunpy/database/tables.py]: line break after the `"""`

File: sunpy/database/tables.py [hunk 7]
```diff
@@ -340,6 +341,66 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
...
+        result block.
+
+        Parameters
+        ----------
+        sr_block : sunpy.net.dataretriever.client.QueryResponseBlock
```
Comment (Cadair) [inline sunpy/database/tables.py]: type should be in backticks

File: sunpy/database/tables.py [hunk 8]
```diff
@@ -340,6 +341,66 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
...
+            A query result block is usually not created directly; instead,
+            one gets instances of
+            ``sunpy.net.dataretriever.client.QueryResponseBlock`` by iterating
+            over each element of a Fido search result.
+        default_waveunit : str, optional
```
Comment (Cadair) [inline sunpy/database/tables.py]: type should be in backticks

File: sunpy/database/tables.py [hunk 9]
```diff
@@ -340,6 +341,66 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
...
+        'lyra'
+
+        """
+        # All attributes of DatabaseEntry that are not in QueryResponseBlock
+        # are set as None for now.
+        source = str(sr_block.source) if sr_block.source is not None else None
+        provider = str(sr_block.provider) if sr_block.provider is not None else None
+        # physobs is written as phyobs in
+        # sunpy.net.dataretriever.client.QueryResponseBlock
+        physobs = getattr(sr_block, 'phyobs', None)
```
Comment (Cadair) [inline sunpy/database/tables.py]: wtaf seriously?!? this needs a separate issue and a patch!

File: sunpy/database/tables.py [hunk 10]
```diff
@@ -340,6 +341,66 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
             instrument=instrument, size=qr_block.size,
             wavemin=wavemin, wavemax=wavemax)
 
+    @classmethod
+    def _from_fido_search_result_block(cls, sr_block, default_waveunit=None):
```
Comment (Cadair) [inline sunpy/database/tables.py]: this method should be named "_from_dataretriever_response_block" Fido is the interface dataretriever is the subpackgae that handles the extra sources.

File: sunpy/database/tables.py [hunk 11]
```diff
@@ -432,6 +493,63 @@ def entries_from_query_result(qr, default_waveunit=None):
         yield DatabaseEntry._from_query_result_block(block, default_waveunit)
 
 
+def entries_from_fido_search_result(sr, default_waveunit=None):
+    """Use a `sunpy.net.dataretriever.downloader_factory.UnifiedResponse`
```
Comment (Cadair) [inline sunpy/database/tables.py]: line break after `"""`

File: sunpy/database/tables.py [hunk 12]
```diff
@@ -432,6 +493,63 @@ def entries_from_query_result(qr, default_waveunit=None):
...
+    object returned from
+    :meth:`sunpy.net.dataretriever.downloader_factory.UnifiedDownloaderFactory.search`
+    to generate instances of :class:`DatabaseEntry`. Return an iterator
+    over those instances.
+
+    Parameters
+    ----------
+    search_result : sunpy.net.dataretriever.downloader_factory.UnifiedResponse
```
Comment (Cadair) [inline sunpy/database/tables.py]: type should be in backticks.

File: sunpy/database/tables.py [hunk 13]
```diff
@@ -432,6 +493,63 @@ def entries_from_query_result(qr, default_waveunit=None):
...
+            A UnifiedResponse object that is used to store responses from the
+            unified downloader. This is returned by the ``search`` method of a
+            :class:`sunpy.net.dataretriever.downloader_factory.UnifiedDownloaderFactory`
+            object.
+
+    default_waveunit : str, optional
```
Comment (Cadair) [inline sunpy/database/tables.py]: type should be in backticks.

File: sunpy/database/tables.py [hunk 14]
```diff
@@ -432,6 +493,63 @@ def entries_from_query_result(qr, default_waveunit=None):
...
+    'http://proba2.oma.be/lyra/data/bsd/2012/01/01/lyra_20120101-000000_lev2_std.fits'
+    >>> entry.observation_time_start, entry.observation_time_end
+    (datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 2, 0, 0))
+    >>> entry.instrument
+    'lyra'
+
+    """
+    for entry in sr:
+        if isinstance(entry, sunpy.net.vso.vso.QueryResponse):
+            #This is because EVE doesn't return a Fido QueryResponse. It
```
Comment (Cadair) [inline sunpy/database/tables.py]: this comment is not quite correct, "this is because Fido can search the VSO".

File: sunpy/database/database.py (hunk 2 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: done

File: sunpy/database/database.py (hunk 3 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: done

File: sunpy/database/database.py (hunk 4 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: done

File: sunpy/database/database.py (hunk 5 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: done

File: sunpy/database/tables.py (hunk 6 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 7 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 8 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 11 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 12 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 13 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 14 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/database.py [hunk 15]
```diff
@@ -857,6 +857,29 @@ def add_from_vso_query_result(self, query_result,
                 query_result, self.default_waveunit),
             ignore_already_added)
 
+    def add_from_fido_search_result(self, search_result,
+                                    ignore_already_added=False):
```
Comment (Cadair) [inline sunpy/database/database.py]: is it my imagination or is this one space out of alignment?

File: sunpy/database/database.py [hunk 16]
```diff
@@ -857,6 +857,29 @@ def add_from_vso_query_result(self, query_result,
...
+        """
+        Generate database entries from a Fido search result and add all the
+        generated entries to this database.
+
+        Parameters
+        ----------
+        search_result : `sunpy.net.dataretriever.downloader_factory.UnifiedResponse`
```
Comment (Cadair) [inline sunpy/database/database.py]: this is not the FQN of this class anymore it's now `sunpy.net.fido_factory.UnifiedResponse`

File: sunpy/database/database.py [hunk 17]
```diff
@@ -857,6 +857,29 @@ def add_from_vso_query_result(self, query_result,
...
+            A UnifiedResponse object that is used to store responses from the
+            unified downloader. This is returned by the ``search`` method of a
+            :class:`sunpy.net.dataretriever.downloader_factory.UnifiedDownloaderFactory`
```
Comment (Cadair) [inline sunpy/database/database.py]: same here.

File: sunpy/database/tables.py [hunk 18]
```diff
@@ -349,6 +350,67 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
...
+        provider = getattr(sr_block, 'provider', None)
+        # physobs is written as phyobs in
+        # sunpy.net.dataretriever.client.QueryResponseBlock
+        physobs = getattr(sr_block, 'phyobs', None)
+        if physobs is not None:
+            physobs = str(physobs)
+        instrument = getattr(sr_block, 'instrument', None)
+        time_start = sr_block.time.start
+        time_end = sr_block.time.end
+        wavemin = None
```
Comment (Cadair) [inline sunpy/database/tables.py]: it is possible wavemin and wavemax exist in the response block. `sunpy.net.dataretriever.client.QueryResponseBlock.wave` is either a quantity (in the case where wavemin == wavemax or it is a tuple of `(wavemin, wavemax)` [iirc there is no client that results in the latter case currently].

File: sunpy/database/tables.py [hunk 19]
```diff
@@ -461,6 +525,62 @@ def entries_from_query_result(qr, default_waveunit=None):
...
+    """
+    Use a `sunpy.net.dataretriever.downloader_factory.UnifiedResponse`
+    object returned from
+    :meth:`sunpy.net.dataretriever.downloader_factory.UnifiedDownloaderFactory.search`
+    to generate instances of :class:`DatabaseEntry`. Return an iterator
+    over those instances.
+
+    Parameters
+    ----------
+    search_result : `sunpy.net.dataretriever.downloader_factory.UnifiedResponse`
```
Comment (Cadair) [inline sunpy/database/tables.py]: and here.

File: sunpy/database/database.py (hunk 15 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: let the pep8 bot decide @Cadair 😒

File: sunpy/database/database.py (hunk 16 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: done

File: sunpy/database/database.py (hunk 17 above)
Comment (Punyaslok) [inline sunpy/database/database.py]: done

File: sunpy/database/tables.py (hunk 18 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py (hunk 19 above)
Comment (Punyaslok) [inline sunpy/database/tables.py]: done

File: sunpy/database/tables.py [hunk 20]
```diff
@@ -349,13 +351,108 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
...
+
+        final_values = {}
+        for key, val in wavelength_temp.items():
+            if isinstance(val, quantity.Quantity):
+                unit = getattr(val, 'unit', None)
+                if unit is None:
+                    if default_waveunit is not None:
+                        unit = Unit(default_waveunit)
+                    else:
+                        raise WaveunitNotFoundError(sr_block)
```
Comment (dpshelio) [inline sunpy/database/tables.py]: Does this mean that in the case of `srs` files this will explode??

File: sunpy/database/tests/test_tables.py [hunk 21]
```diff
@@ -96,6 +113,108 @@ def test_tag_hashability():
...
+        wavemin=np.nan, wavemax=np.nan,
+        instrument='lyra')
+    # 54 entries from EVE
+    assert entries[2] == DatabaseEntry(
+        source='SDO', provider='LASP', physobs='irradiance',
+        fileid='EVE_L1_esp_2012001_00',
+        observation_time_start=datetime(2012, 1, 1, 0, 0),
+        observation_time_end=datetime(2012, 1, 2, 0, 0),
+        instrument='EVE', size=-1.0,
+        wavemin=0.1, wavemax=30.400000000000002)
```
Comment (dpshelio) [inline sunpy/database/tests/test_tables.py]: really??? 30.400000000000000002 ??? Should we keep the rounding that was in the wavelength comparison?

File: sunpy/database/tests/test_tables.py [hunk 22]
```diff
@@ -96,6 +113,108 @@ def test_tag_hashability():
...
+        wavemin=np.nan, wavemax=np.nan,
+        instrument='noaa-predict')
+    # 2 entries from norh
+    assert entries[60] == DatabaseEntry(
+        source='NAOJ', provider='NRO', physobs="",
+        fileid=("ftp://anonymous:data@sunpy.org@solar-pub.nao.ac.jp/"
+                "pub/nsro/norh/data/tcx/2012/01/tca120101"),
+        observation_time_start=datetime(2012, 1, 1, 0, 0),
+        observation_time_end=datetime(2012, 1, 2, 0, 0),
+        wavemin=17634850.470588233, wavemax=17634850.470588233,
```
Comment (dpshelio) [inline sunpy/database/tests/test_tables.py]: This cannot be right... Do we have even that precision in wavelength?

File: sunpy/database/tables.py [hunk 23]
```diff
@@ -349,13 +351,108 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
...
+            instrument=instrument, size=size,
+            wavemin=wavemin, wavemax=wavemax)
+
     def __eq__(self, other):
-        wavemins_equal = self.wavemin is None and other.wavemin is None or\
-            self.wavemin is not None and other.wavemin is not None and\
-            round(self.wavemin, 10) == round(other.wavemin, 10)
-        wavemaxs_equal = self.wavemax is None and other.wavemax is None or\
-            self.wavemax is not None and other.wavemax is not None and\
-            round(self.wavemax, 10) == round(other.wavemax, 10)
```
Comment (dpshelio) [inline sunpy/database/tables.py]: I believe we should keep the `round` for comparisons.

Comment (Cadair) [inline sunpy/database/tables.py]: we were rounding to 10 d.p.! `all_close` is much more appropriate for this, if you want to reduce the precision of the comparison changing `rtol` in `all_close` is the correct way to proceed.

File: sunpy/database/tests/test_tables.py (hunk 21 above)
Comment (Cadair) [inline sunpy/database/tests/test_tables.py]: yeah we probably don't need that, it's just floating point noise at that stage.

File: sunpy/database/tests/test_tables.py (hunk 22 above)
Comment (Cadair) [inline sunpy/database/tests/test_tables.py]: it's just number comparison, it's not saying anything about the accuracy of the number.

Comment (Cadair) [inline sunpy/database/tests/test_tables.py]: I think these numbers have gone through a round of GHz -> nm conversion, which is probably the source of the precision.

File: sunpy/database/tables.py (hunk 20 above)
Comment (Cadair) [inline sunpy/database/tables.py]: I have no idea what would happen if you tried to put an SRS table in this, it was built for FITS files, I think that is out of scope for this PR.
